### DISCUSSION TITLE
Parsing in Bitwuzla

### DISCUSSION THREAD

**Discussion by daniel-raffler** [bitwuzla/bitwuzla#101] [General] [answered]
Hello everyone,
We'd like to use Bitwuzla as one of our backends in JavaSMT and one of the issue that is holding is back is parser support. Bitwuzla does support reading in SMTLIB2 files through its API, however, this always seems to spawn a new solver instance and there is no way to carry over results between different parser runs. Here's a quick example to illustrate the point:
```
#include <bitwuzla/cpp/bitwuzla.h>
#include <bitwuzla/cpp/parser.h>

#include <string>
#include <fstream>

#include <cassert>

using namespace bitwuzla;
using namespace bitwuzla::parser;

/** Write string to a file, then run the bitwuzla parser and return the assertion as a bitwuzla term. */
Term parseString(std::string formula) {
  std::string path = std::tmpnam(nullptr);

  std::ofstream myfile;
  myfile.open(path);
  myfile << formula;
  myfile.close();

  Options options;
  options.set(Option::PRODUCE_MODELS, (uint64_t) 2);
  options.set(Option::REWRITE_LEVEL, (uint64_t) 0);
  
  Parser parser(options, path);
  std::string err = parser.parse(true);

  assert(err.empty()); // We don't expect any parsing errors

  Bitwuzla *bitwuzla = parser.bitwuzla().get();
  return bitwuzla->get_assertions()[0];
}

int main() {
  Term a_true  = parseString("(declare-const a Bool) (assert (= a true))");
  Term a_false = parseString("(declare-const a Bool) (assert (= a false))");

  Bitwuzla bitwuzla;
  bitwuzla.assert_formula(a_true);
  bitwuzla.assert_formula(a_false);

  Result is_sat = bitwuzla.check_sat();
  assert(is_sat == Result::UNSAT);
}
```

This example reads in two simple SMTLIB2 files. The first asserts that "a" is true, and the second asserts that "a" is false. If we then get the assertions from the parser and push them onto our solver stack we expect the result to always be unsatisfiable. Hoewever, in Bitwuzla the result is SAT as the "a" variables from different parser runs are not the same.

This causes some issues for us as we would like to use the parser to read in short formulas as SMTLIB2 scripts. This is needed, for instance, to translate formulas between different solvers by first printing them, and then parsing them again. For this we don't need the full SMTLIB2 language. Our formulas always consists of several (declare-fun ...) followed by a single assertion. All variables in the assertion must be declared, but this may also happen earlier through the API and variables should be "persistent" across multiple parser runs.

Would it be possible to do something like this in Bitwuzla?

Thanks,
Daniel

**Comment by aniemetz** [comment]
Hi Daniel, the problem here is that you create a new parser instance in each call to `parseString`. You must use the same parser to parser input to make sure that the symbol table is in sync, otherwise a symbol `a` will map to 2 different terms. Now, the next issue is that currently, a parser is tied to an input file and doesn't allow to parser from a string/stream yet. This is on our TODO list (#79). We will address this shortly.

**Comment by daniel-raffler** [comment]
Hello Aina,
thanks for your help! Great to hear that you are working on this.

**Comment by aniemetz** [comment]
Parsing from string is now supported on main. See https://github.com/bitwuzla/bitwuzla/blob/main/examples/c/parse.c, https://github.com/bitwuzla/bitwuzla/blob/main/examples/cpp/parse.cpp and https://github.com/bitwuzla/bitwuzla/blob/main/examples/python/parse.py for examples. Note that the C++ API also allows to parse from an input stream.

**Comment by daniel-raffler** [comment]
Hello Aina,
thanks for your work! This should already help simplify our code quite a bit. There are two more issues that I'd like to ask about: The first is whether it would be possible to add a method to the parser that returns the current symbol table. Something like `map<string, variant<Sort, Term>> get_defined_symbols()` where all the symbols that were defined in the input file can be found. The reason here is that JavaSMT uses its own cache to avoid creating duplicate variables and we'd like to keep this synchronized with the symbols from the parser. Otherwise we run into problems when trying to mix code that uses the API and code that uses the parser:
```
  public void parseDeclareAfterQueryTest1() {
    BooleanFormula formula = mgr.parse("(declare-fun var () Bool)(assert var)");
    BooleanFormula var = bmgr.makeVariable("var");
    Truth.assertThat(mgr.extractVariables(formula).values()).containsExactly(var);
  }
```
In this test case two distinct variable (both named "var") are created.  Usually `makeVariable` would avoid this by checking its internal cache first, but it has no way of knowing about variables that were defined by the parser.

The other issue is that the Bitwuzla parser does not allow variables to be redeclared. This can cause problems when reusing the parser:
```
 public void parseTwiceTest1() {
    BooleanFormula formula1 = mgr.parse( "(declare-fun x () Bool)(assert x)";);
    BooleanFormula formula2 = mgr.parse( "(declare-fun x () Bool)(assert x)";);
    Truth.assertThat(formula1).isEqualTo(formula2);
  }
```
Here the second run fails as the variable "x" has already been declared in the first run. Would it be possible to allow such redeclarations as long as the sort does not change?

Thanks again,
Daniel

**Comment by aniemetz** [comment]
@daniel-raffler you can retrieve terms that are created via `parse("(declare-fun var () Bool)(assert var)")` with `parse_term("var")` (see https://github.com/bitwuzla/bitwuzla/blob/main/examples/cpp/parse.cpp). However, you cannot use this to "test" if a symbol with that name already exists (it will trigger a parse error). It might make sense to provide an API function to export defined sorts/terms, we'll put it on our TODO list and will keep you posted.

As for your second question, the parser is our textual interface and as such must comply with SMT-LIB. Your input is not SMT-LIB compliant and will thus fail. We will not change this behavior. I also don't see the need for redeclaring symbols, see my comment above.

**Comment by daniel-raffler** [comment]
Hello Aina,
thanks for your reply! A function to export defined sort/terms would already be quite useful for us.

**Comment by aniemetz** [comment, marked as answer]
@daniel-raffler We added support for `Parser::get_declared_sorts()` and `Parser::get_declared_funs()`. Note that these two functions return sorts declared via `declare-sort` and funs declared via `declare-const` and `declare-fun` as a vector for Sort resp. Term. The choice to return these as a vector is on purpose, to also be able to handle potential future SMT-LIB scenarios where shadowing is allowed (outside of what is currently already allowed with binders).

**Comment by daniel-raffler** [comment]
Hello Aina,
thanks for your great work! This should really help us out.
